force-app/main/default/classes/DMLOperations.cls

```diff
@@ -181,14 +229,16 @@ public with sharing class DMLOperations {
 	 */
 	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
 		// Loop through each Contact
-
+        for (Contact con : contactsToUpsert) {
 		// Extract the Account Name from Contact's LastName
-
+            String accountName = con.LastName;
 		// Get the account based on the name or create one if it doesn't exist
-
+            Account accountToLink = upsertAccount(accountName);
 		// Link the Contact to its Account using the Account Id
-
+            con.AccountId = accountToLink.Id;
+             }
```

Comment:
Look at the way you solved upsertOpportunities. You can also use maps + single queries here to avoid doing SOQL + DML in the for loop. `upsertAccount` is the issue here.